fix(ivy): match attribute selectors for content projection with inline-templates

petebacondarwin opened the pull request:
See FW-1002

Supercedes #28655

kara commented on packages/core/src/render3/node_selector_matcher.ts:
```diff
@@ -154,33 +168,78 @@ function readClassValueFromTNode(tNode: TNode): string {
+    }
+  } else {
+    // Check the Template attributes
+    i = attrs.indexOf(AttributeMarker.Template);
```
Break this out into a helper function for readability?  `getIndexInTemplateAttrs` or `searchTemplateAttrs` ?

kara reviewed (changes requested):
Approach looks great, just minor comments

petebacondarwin commented on packages/core/src/render3/di.ts:
```diff
@@ -272,7 +272,9 @@ export function injectAttributeImpl(tNode: TNode, attrNameToInject: string): str
-      if (attrName === AttributeMarker.SelectOnly) break;
+      if (isNameOnlyAttributeMarker(attrName)) break;
+      // TODO: What happens here if an attribute has a namespace?
+      // TODO: What happens here if the attribute name happens to match a CSS class or style?
```
I think that the logic is also wrong for classes and styles considering the comment here https://github.com/angular/angular/blob/84406e4d6d93b28b23efbb1701bc5ae1084da67b/packages/core/src/render3/instructions.ts#L1477-L1481

This logic assumes that attributes (until the "name-only" attributes) always come in name-value pairs. But classes and styles do not follow this rule. E.g. `CLASSES_MARKER, class1, class2, class3,`.

Therefore, I was thinking that we should also break when when we hit one of `Classes` or `Styles` markers.

petebacondarwin replied:
Or are you saying that by the time we get here any `Classes` or `Styles` attributes will have been removed from the array?

kara replied:
Oh, you're totally right. I had forgotten that class and style attributes only contain name data, so it would be a larger effort to get the values out of the styling context. Created a Jira for that too (https://angular-team.atlassian.net/browse/FW-1139).

```suggestion
      // TODO(FW-1139): support classes / styles in @Attribute injection
```

petebacondarwin commented on packages/core/src/render3/node_selector_matcher.ts:
```diff
@@ -154,33 +168,78 @@ function readClassValueFromTNode(tNode: TNode): string {
+      do {
+        i++;
+        const maybeAttrName = attrs[i];
+        if (maybeAttrName === name) return i;
```
I did originally try to reuse the `selectOnlyMode` approach but this code is more complex now that we have to deal with `Template` markers and also different code paths depending upon the `projectionMode`. I will see what I can do to reduce the code size...

petebacondarwin replied:
Thanks for pushing back on this @kara  - I think I have come up with a much smaller and cleaner block of code.

JoostK commented on packages/core/src/render3/node_selector_matcher.ts:
```diff
@@ -288,3 +267,26 @@ export function matchingProjectionSelectorIndex(
+
+function getNameOnlyMarkerIndex(nodeAttrs: TAttributes) {
+  let nameOnlyMarkerIdx = nodeAttrs && nodeAttrs.length;
+  for (let i = 0; i < nodeAttrs.length; i++) {
```
You test `nodeAttrs` being truthy above, then dereference it here unguarded. Given the parameter has a non-nullable type, the truth check above is probably superfluous.

**edit**: I now see this is existing code.

petebacondarwin replied:
I'll fix

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -7,6 +7,7 @@
  */
 
 import {AST, BindingType, BoundTarget, ImplicitReceiver, PropertyRead, TmplAstBoundText, TmplAstElement, TmplAstNode, TmplAstTemplate, TmplAstVariable} from '@angular/compiler';
+import {BoundAttribute, TextAttribute} from '@angular/compiler/src/render3/r3_ast';
```
I try to avoid deep imports - feel free to add these types to the exports of `@angular/compiler` if needed.

petebacondarwin replied:
Auto-import fail :-)

petebacondarwin replied:
These are not even being used in this file!

petebacondarwin replied:
Agh! I added them back in!! I'll fix.

JoostK commented on packages/core/src/render3/node_selector_matcher.ts:
```diff
@@ -288,3 +267,26 @@ export function matchingProjectionSelectorIndex(
+  for (let i = 0; i < nodeAttrs.length; i++) {
+    const nodeAttr = nodeAttrs[i];
+    if (isNameOnlyAttributeMarker(nodeAttr)) {
+      nameOnlyMarkerIdx = i;
```
Was it deliberate to assign and break, instead of returning at this point? We could get rid of `nameOnlyMarkerIdx` that way now that has been broken out into its own function.

petebacondarwin replied:
No that was a mistake when I extracted this code into a function from the main body of the code.

JoostK commented on packages/core/src/render3/node_selector_matcher.ts:
```diff
@@ -288,3 +267,26 @@ export function matchingProjectionSelectorIndex(
+
+function matchTemplateAttribute(attrs: TAttributes, name: string): number {
+  let i = attrs.indexOf(AttributeMarker.Template);
+  if (i > -1 && i < attrs.length) {
```
Looks like the `do-while` could become a `while`, as the do-block is guarded within the exact same condition as the while guard:

```suggestion
  if (i !== -1) {
    while (i < attrs.length) {
      i++;
      if (attrs[i] === name) return i;
    }
  }
```

petebacondarwin replied:
It's not quite trivial but I think that we can do:

```ts
function matchTemplateAttribute(attrs: TAttributes, name: string): number {
  let i = attrs.indexOf(AttributeMarker.Template);
  if (i > -1) {
    i++;
    while (i < attrs.length) {
      if (attrs[i] === name) return i;
      i++;
    }
  }
  return -1;
}
```

JoostK replied:
Oh I misjudged the `i++`! Looking at your proposal, we could replace the 2 `i++`'s with `while (++i < attrs.length)` (or would it then become *clever code* 🙄)

petebacondarwin replied:
I'm not going there :-P

JoostK commented on packages/core/src/render3/util/attrs_utils.ts:
```diff
@@ -105,3 +105,14 @@ export function attrsStylingIndexOf(attrs: TAttributes, startIndex: number): num
 }
+
+/**
+ * Test whether a the given value is a marker that indicates that the following
```
```suggestion
 * Test whether the given value is a marker that indicates that the following
```

JoostK commented on packages/core/src/render3/node_selector_matcher.ts:
```diff
@@ -66,14 +67,7 @@ export function isNodeMatchingSelector(
-      break;
-    }
-  }
+  let nameOnlyMarkerIdx = getNameOnlyMarkerIndex(nodeAttrs);
```
```suggestion
  const nameOnlyMarkerIdx = getNameOnlyMarkerIndex(nodeAttrs);
```

alxhub commented on packages/compiler/src/render3/view/t2_binder.ts:
```diff
@@ -380,7 +381,7 @@ class TemplateBinder extends RecursiveAstVisitor implements Visitor {
     if (template instanceof Template) {
       // For <ng-template>s, process inputs, outputs, variables, and child nodes. References were
```
Update the comment :-)

alxhub commented on packages/compiler/src/render3/view/t2_binder.ts:
```diff
@@ -402,7 +403,7 @@ class TemplateBinder extends RecursiveAstVisitor implements Visitor {
 
   visitTemplate(template: Template) {
     // First, visit the inputs, outputs of the template node.
```
And here

alxhub commented on packages/compiler/src/render3/r3_ast.ts:
```diff
@@ -79,12 +79,18 @@ export class Element implements Node {
+   * These are used by the template TypeChecker for working out the scope of expressions
+   * in templates. See `TemplateBinder`.
+   */
+  public boundAttributes = this.inputs.concat(this.templateAttrs.filter(isBoundAttribute));
```
I don't think it's a good idea to include this "summary" computed field on the AST. Instead, we should modify `t2_binder` to understand that `Template` nodes have both `inputs` and `templateAttrs` that need to be processed.

petebacondarwin replied:
This `boundAttributes` property is used twice in the `t2_binder` but also in the `type_check_block`. Computing it is not trivial so I thought it was best to compute it only once per template node and cache it there.

petebacondarwin replied:
Discussed offline and I am going to refactor :-)

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -430,7 +431,8 @@ function tcbProcessTemplateDeclaration(tmpl: TmplAstTemplate, tcb: Context, scop
         // For each template guard function on the directive, look for a binding to that input.
-        const boundInput = tmpl.inputs.find(i => i.name === inputName);
+        const boundInput = tmpl.inputs.find(i => i.name === inputName) ||
+            tmpl.templateAttrs.filter(isBoundAttribute).find(i => i.name === inputName);
```
```suggestion
            tmpl.templateAttrs.find((i: TextAttribute|BoundAttribute): i is BoundAttribute => i instanceof BoundAttribute && i.name === inputName);
```

This avoids the cost of `filter()` for every lookup while still narrowing the type correctly.

alxhub commented on packages/compiler/src/render3/view/t2_binder.ts:
```diff
@@ -285,14 +285,15 @@ class DirectiveBinder<DirectiveT extends DirectiveMeta> implements Visitor {
-        this.bindings.set(binding, node);
-      }
-    });
+    [...node.attributes, ...(node instanceof Template ? node.templateAttrs : node.inputs)].forEach(
```
For templates, shouldn't we iterate over `inputs` as well as `templateAttrs`?

kara reviewed (approved):
LGTM

petebacondarwin committed:
fix(compiler): ensure template is updated if an output is transformed

No idea where the tests for this code are, but it looks like this was an
oversight in the original commit 6a0f78.

petebacondarwin committed:
refactor: rename `matchingSelectorIndex` to `matchingProjectionSelectorIndex`

The previous name was ambiguous as there are different strategies
for matching selectors depending upon the scenario.

petebacondarwin committed:
refactor(ivy): define new `AttributeMarker.Template` marker

This commit adds a new `AttributeMarker` type that will be used, in a
future commit, to mark attributes as coming from an inline-template
expansion, rather than the element that is being contained in the template.

petebacondarwin committed:
refactor: rename `AttributeMarker.ProjectOnly` to `AttributeMarker.Bindings`

petebacondarwin committed:
fix(ivy): teach template type checker about template attributes

For the template type checking to work correctly, it needs to know
what attributes are bound to expressions or directives, which may
require expressions in the template to be evaluated in a different
scope.

In inline templates, there are attributes that are now marked as
"Template" attributes. We need to ensure that the template
type checking code looks at these "bound" attributes as well as the
"input" attributes.

petebacondarwin committed:
fix(ivy): match attribute selectors for content projection with inline-templates

The content projection mechanism is static, in that it only looks at the static
template nodes before directives are matched and change detection is run.
When you have a selector-based content projection the selection is based
on nodes that are available in the template.

For example:

```
<ng-content selector="[some-attr]"></ng-content>
```

would match

```
<div some-attr="..."></div>
```

If you have an inline-template in your projected nodes. For example:

```
<div *ngIf="..." some-attr="..."></div>
```

This gets pre-parsed and converted to a canonical form.

For example:

```
<ng-template [ngIf]="...">
  <div some-attr=".."></div>
</ng-template>
```

Note that only structural attributes (e.g. `*ngIf`) stay with the `<ng-template>`
node. The other attributes move to the contained element inside the template.

When this happens in ivy, the ng-template content is removed
from the component template function and is compiled into its own
template function. But this means that the information about the
attributes that were on the content are lost and the projection
selection mechanism is unable to match the original
`<div *ngIf="..." some-attr>`.

This commit adds support for this in ivy. Attributes are separated into three
groups (Bindings, Templates and "other"). For inline-templates the Bindings
and "other" types are hoisted back from the contained node to the `template()`
instruction, so that they can be used in content projection matching.

petebacondarwin committed:
test(ivy): add tests for projection on inline-templates

alxhub commented:
[Presubmit](http://test/OCL:237271739:BASE:237271776:1551982678490:f79525f)
